src/core/Action.js

```diff
@@ -0,0 +1,224 @@
+          resultCallback(message.values);
+        }
+      });
+    }
+
+    var call = {
+      op : 'send_action_goal',
+      id : actionGoalId,
+      action : this.name,
+      action_type: this.actionType,
+      args : goal,
+      feedback : true,
+    };
+    this.ros.callOnConnection(call);
+
+    return actionGoalId;
+  }
+
+  /**
+   * Cancels an action goal.
+   * 
+   * @param {string} id - The ID of the action goal to cancel.
+   */
+  cancelGoal(id) {
+    var call = {
+      op: 'cancel_action_goal',
+      id: id,
+      action: this.name,
+    };
+    this.ros.callOnConnection(call);
+  }
+
+  /**
+   * @callback advertiseActionCallback
+   * @param {TGoal} goal - The action goal.
+   * @param {string} id - The ID of the action goal to execute.
+   */
+  /**
+   * @callback advertiseCancelCallback
+   * @param {string} id - The ID of the action goal to cancel.
+   */
+  /**
+   * Advertise the action. This turns the Action object from a client
+   * into a server. The callback will be called with every goal sent to this action.
+   *
+   * @param {advertiseActionCallback} actionCallback - This works similarly to the callback for a C++ action.
+   * @param {advertiseCancelCallback} cancelCallback - A callback function to execute when the action is canceled.
+   */
+  advertise(actionCallback, cancelCallback) {
+    if (this.isAdvertised || typeof actionCallback !== 'function') {
+      return;
+    }
+
+    this._actionCallback = actionCallback;
+    this._cancelCallback = cancelCallback;
+    this.ros.on(this.name, this._executeAction.bind(this));
+    this.ros.callOnConnection({
+      op: 'advertise_action',
+      type: this.actionType,
+      action: this.name
+    });
+    this.isAdvertised = true;
+  }
+
+  /**
+   * Unadvertise a previously advertised action.
+   */
+  unadvertise() {
+    if (!this.isAdvertised) {
+      return;
+    }
+    this.ros.callOnConnection({
+      op: 'unadvertise_action',
+      action: this.name
+    });
+    this.isAdvertised = false;
+  }
+
+  /**
+   * Helper function that executes an action by calling the provided
+   * action callback with the auto-generated ID as a user-accessible input.
+   * Should not be called manually.
+   * 
+   * @param {Object} rosbridgeRequest - The ROSLIB.ActionGoal to send.
+   * @param {string} rosbridgeRequest.id - The ID of the action goal.
+   */
+  _executeAction(rosbridgeRequest) {
+    var id;
+    if (rosbridgeRequest.id) {
+      id = rosbridgeRequest.id;
+    }
+
+    // If a cancellation callback exists, call it when a cancellation event is emitted.
+    if (typeof id === 'string') {
+      this.ros.on(id, (message) => {
+        if (message.op === 'cancel_action_goal' && typeof this._cancelCallback === 'function') {
+          this._cancelCallback(id);
+        }
+      });
+    }
+
+    // Call the goal execution function provided.
+    // @ts-expect-error -- possibly null
+    this._actionCallback(rosbridgeRequest.args, id);
```

Comment:
Let's try not to add any more typescript workarounds now that we're adding new functionality. Does this work?

```suggestion
    if (this._actionCallback !== null) {
      this._actionCallback(rosbridgeRequest.args, id);
    }
```

Reply:
Good call, to be consistent with other code in this library I checked that its type was `'function'` rather than null, and also patched up a few more things related to stripping away this comment. Let me know!